protocol/codec.go

```diff
@@ -47,6 +47,11 @@ type Decoder interface {
 	Decode(objptr interface{}) error
 }
 
+// MsgpDecoder is our interface for a thing that can msgp-decode objects.
+type MsgpDecoder interface {
+	Decode(objptr msgp.Unmarshaler) error
+}
+
```

Comment:
I don't think we should introduce an interface.  Just return the `*MsgpDecoder` in `NewMsgpDecoderBytes`.